### DISCUSSION TITLE
데이터 압축하면 어떤 이점이 있을까?   (feat.레디스)

### DISCUSSION THREAD

**Discussion by devjun10** [dailyge/dailyge-server#86] [Performance]
## 📝 연관 PR

- [[DLG-131] 레디스에 저장되는 데이터를 압축한다.](https://github.com/dailyge/dailyge-server/pull/70)
- [[DLG-253] GZip 압축을 ZStandard로 변경한다.](https://github.com/dailyge/dailyge-server/pull/138)

<br/><br/>

## ❓ 작성 이유

레디스에 데이터를 저장할 때, 압축을 하면 어떤 이점이 있는지 공유하기 위해 글을 작성하게 되었습니다.

<br/><br/>

## 📝 내용

데이터를 압축하면 **`저장 용량 증가`**, **`데이터 전송 속도 증가`**, **`네트워크 대역폭을 효율적으로 사용`** 할 수 있습니다.

1. 저장 용량 증가
2. 데이터 전송 속도 증가
3. 네트워크 대역폭 효율적 사용

<br/><br/><br/><br/>

## 1. 저장 용량 증가

데이터를 압축하면 **`저장 용량을 증가`** 시킬 수 있습니다. 아래는 사용자 정보 캐시를 압축한 것인데, **`약 13%`** 정도 압축됐죠. 이는 데이터의 크기가 클 수록 더 큰 효과를 보게 됩니다. 즉, 이를 통해 데이터를 저장할 때, 저장 공간을 더 확보할 수 있습니다.

```java
@DisplayName("[UnitTest] 압축 단위 테스트")
class CompressionUnitTest {

    ......

    @Test
    @DisplayName("객체 데이터를 압축하면 사이즈가 작아진다.")
    void whenCompressObjectThenDataSizeShouldBeSmaller() {
        final byte[] originalData = serialize(userCache);
        final byte[] compressedData = compressAsByteArray(userCache);

        assertTrue(compressedData.length < originalData.length);

        log.info("Original Data Size: {}", originalData.length);
        log.info("Compressed Data Size: {}", compressedData.length);
    }

    ......

}
```

```shell
[main] INFO project.dailyge.test.common.CompressionUnitTest -- Original Data Size: 311
[main] INFO project.dailyge.test.common.CompressionUnitTest -- Compressed Data Size: 270
```

<br/><br/><br/><br/>

현재 사용중인 레디스 서버의 용량이 작지만, 부하 테스트를 할 때, 30분 간 300만 건 이상의 데이터가 저장되어도 큰 문제가 없는 것을 봤습니다. 물론 이후엔 레디스가 뻗었지만요😡 여튼 데이터를 압축하면 **`더 많은 데이터를 저장`** 할 수 있는 장점이 있습니다

![image](https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2F9OZ3M%2FbtsJdbko2C8%2FzoYm0WTqbu1Db6QaKgOCUk%2Fimg.png)

<br/><br/><br/><br/>

물론 무조건 압축한다고 좋은 것은 아닌데, 압축 과정에서 메타 데이터가 필요하기 때문입니다. 너무 작은 데이터를 압축하면 역으로 압축 오버헤드가 발생해 데이터 크기가 증가하게 됩니다. 따라서 어느정도 크기가 있는 데이터를 압축해 성능상 이점을 보도록 합니다.

```java
@DisplayName("[UnitTest] 압축 단위 테스트")
class CompressionUnitTest {

    ......

    @Test
    @DisplayName("너무 작은 데이터를 압축하면, 메타 데이터로 인해 데이터 크기가 오히려 커진다.")
    void whenCompressTooSmallDataThenSizeShouldBeBiggerThanOriginalData() {
        final Long userId = 1L;
        final byte[] originalData = serialize(userId);
        final byte[] compressedData = compressAsByteArray(userId);

        assertTrue(compressedData.length > originalData.length);

        log.info("Original Data Size: {}", originalData.length);
        log.info("Compressed Data Size: {}", compressedData.length);
    }

    ......

}
```

```shell
[main] INFO project.dailyge.test.common.CompressionUnitTest -- Original Data Size: 82
[main] INFO project.dailyge.test.common.CompressionUnitTest -- Compressed Data Size: 90
```

<br/><br/><br/><br/>

## 2. 데이터 전송 속도 증가

데이터의 크기가 작아지면, 네트워크를 탈 때, 더 작은 데이터가 전송되므로, 전송 속도도 빨라지겠죠? 

> 이 부분에 대한 자세한 설명은 생략하도록 하겠습니다.

<br/><br/><br/><br/>

## 3. 네트워크 대역폭 효율적 사용

네트워크 대역폭은 **`시간 당 전송할 수 있는 최대 데이터의 양`** 입니다. AWS의 [공식 문서](https://docs.aws.amazon.com/ko_kr/AWSEC2/latest/UserGuide/ec2-instance-network-bandwidth.html)를 보면 대역폭에 대한 내용이 나와있습니다. 즉, 데이터를 압축 시키면 단위 시간 당 전송할 수 있는 데이터가 많아지므로, 대역폭을 효율적으로 사용할 수 있게 됩니다.

> 저희 프로젝트는 작아서 이 이점까지 가져가진 못하겠지만, 이런 내용도 있다는 것을 공유하기 위해 작성했습니다.

**Comment by beatmeJY** [comment]
너무 유익한 정보 감사합니다!!😄
이러한 저장되는 크기를 줄이는 것만으로도 성능적으로나 서버비용측으로도 많은 이득이 되겠네요!👍🏻

**Comment by kmularise** [comment]
데이터를 압축시키면 동일한 용량으로 더 많은 데이터들을 보낼 수 있어서 저장용량 뿐만 아니라 데이터 전송 속도 측면에서도 이점을 얻을 수 있군요! 추가로 일정 크기 이상의 데이터만 압축시킬 때, 이점을 보는 것을 확인했습니다!
